OrbitCaptureClient/CaptureEventProcessor.cpp

```diff
@@ -340,3 +428,254 @@ void CaptureEventProcessor::SendTracepointInfoToListenerIfNecessary(
+
+  for (const auto& completed_marker : gpu_queue_submission.completed_markers()) {
+    CHECK(first_command_buffer != std::nullopt);
+    TimerInfo marker_timer;
+
+    // If we've recorded the submission that contains the begin marker, we'll retrieve this
+    // submission from our mappings, and set the markers begin time accordingly.
+    // Otherwise, we will use the capture start time as begin.
+    if (completed_marker.has_begin_marker()) {
+      const auto& begin_marker_info = completed_marker.begin_marker();
+      const auto& begin_marker_meta_info = begin_marker_info.meta_info();
+      const int32_t begin_marker_thread_id = begin_marker_meta_info.tid();
+      uint64_t begin_marker_post_submission_cpu_timestamp =
+          begin_marker_meta_info.post_submission_cpu_timestamp();
+      uint64_t begin_marker_pre_submission_cpu_timestamp =
+          begin_marker_meta_info.pre_submission_cpu_timestamp();
+
+      const GpuQueueSubmission* matching_begin_submission = nullptr;
+      if (submission_pre_submission_cpu_timestamp == begin_marker_pre_submission_cpu_timestamp &&
+          submission_post_submission_cpu_timestamp == begin_marker_post_submission_cpu_timestamp &&
+          submission_thread_id == begin_marker_thread_id) {
+        matching_begin_submission = &gpu_queue_submission;
+      } else {
+        matching_begin_submission = FindMatchingGpuQueueSubmission(
+            begin_marker_thread_id, begin_marker_post_submission_cpu_timestamp);
+      }
+
+      // Note, we receive submissions of a single queue in order (by CPU submission time). So if
+      // there is no matching "begin submission", the "begin" was submitted before the "end" and
+      // we lost the record of the "begin's" submission (which should not happen).
+      CHECK(matching_begin_submission != nullptr);
+
+      std::optional<GpuCommandBuffer> begin_submission_first_command_buffer =
+          FindFirstCommandBuffer(*matching_begin_submission);
+      CHECK(begin_submission_first_command_buffer.has_value());
+
+      const GpuJob* matching_begin_job = FindMatchingGpuJob(
+          begin_marker_thread_id, begin_marker_meta_info.pre_submission_cpu_timestamp(),
+          begin_marker_post_submission_cpu_timestamp);
+      CHECK(matching_begin_job != nullptr);
+
+      marker_timer.set_start(completed_marker.begin_marker().gpu_timestamp_ns() +
+                             matching_begin_job->gpu_hardware_start_time_ns() -
+                             begin_submission_first_command_buffer->begin_gpu_timestamp_ns());
+      if (begin_marker_thread_id == gpu_queue_submission.meta_info().tid()) {
+        marker_timer.set_thread_id(begin_marker_thread_id);
+      }
+
+      DecrementUnprocessedBeginMarkers(begin_marker_thread_id,
+                                       begin_marker_post_submission_cpu_timestamp);
+    } else {
+      marker_timer.set_start(begin_capture_time_ns_);
+      marker_timer.set_thread_id(-1);
```

Comment:
So the thread id is lost, it would only come with the begin marker?

And should -1 be a constant somehow? It would probably be complicated to propagate it to the entire codebase though. I guess it should be a constant defined in the `TimerInfo` proto but it seems that that can't be done. Hmm.

Reply:
I would at least add a local constant to give a name to the `-1` here.

> It does not need to be lost. Basically the point is that (if the debug marker is begins and ends on a different submission), the thread id of the "begin submission" may be different than the one of the "end submission".

I'm not sure how this applies here. Isn't this code executed when `if (completed_marker.has_begin_marker())` is false? Doesn't that just mean that the begin was missing? Or does it mean it was in a different submission?

Reply:
You are totally right. That branch was only if we have not recorded the begin marker. The other case was missing (I added it a couple of lines above). Also added a local constant.